File: Algorithms/126.WordLadderII.py
class Solution(object):
    def findLadders(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: List[List[str]]
        """
        wordDict = set(wordList)
        self.result, self.path, self.parents, levels = [], [], {}, {}

        if endWord not in wordDict:
            return self.result
        wordDict.remove(endWord)

        q = [beginWord]
        levels[beginWord] = 1
        level, found = 0, False
        while q and not found:
            level += 1
            for j in range(len(q)):
                tmp = q.pop(0)
                for i in range(len(tmp)):
                    for c in string.lowercase:
                        if c == tmp[i]:
                            continue
                        temp = tmp[:i] + c + tmp[i+1:]
                        if temp == endWord:
                            self.parents.get(endWord, []).append(tmp)
                            found = True
                        elif temp in levels and level < levels[temp]:
                            self.parents.get(temp).append(tmp)
                        if temp not in wordDict:
                            continue
                        wordDict.remove(temp)
                        q.append(temp)
                        levels[temp] = levels[tmp] + 1
                        self.parents[temp] = [tmp]
        if found:
            self.path.append(endWord)
            self.buildPaths(endWord, beginWord)
        return self.result

    def buildPaths(self, word, beginWord):
        if word == beginWord:
            self.result.append(self.path[:])
            return
        for w in self.parents[word]:
            self.path.append(w)
            self.buildPaths(w, beginWord)
            self.path.pop(-1)

File: Algorithms/java/126.WordLadderII.java
class Solution {
    public List<List<String>> findLadders(String beginWord, String endWord, List<String> wordList) {
        Set<String> dict = new HashSet<>();
        for (String word: wordList)
            dict.add(word);

        List<List<String>> result = new ArrayList<List<String>>();
        if (!dict.contains(endWord))
            return result;
        dict.remove(endWord);

        Queue<String> q = new ArrayDeque<>();
        q.offer(beginWord);

        Map<String, List<String>> parents = new HashMap<String, List<String>>();
        Map<String, Integer> levels = new HashMap<>();
        levels.put(beginWord, 1);
        int level = 0;
        boolean found = false;
        while (!q.isEmpty() && !found) {
            level ++;
            for (int size=q.size(); size > 0; size--) {
                String tmp = q.poll();
                char[] chs = tmp.toCharArray();
                for (int i=0; i < tmp.length(); i++) {
                    char ch = chs[i];
                    for (char c='a'; c <= 'z'; c++) {
                        if (c == ch)
                            continue;
                        chs[i] = c;
                        String temp = new String(chs);
                        if (temp.equals(endWord)) {
                            if (parents.containsKey(endWord)) {
                                parents.get(endWord).add(tmp);
                            } else {
                                List<String> p = new ArrayList<>();
                                p.add(tmp);
                                parents.put(endWord, p);
                            }
                            found = true;
                        } else if (levels.containsKey(temp) && level < levels.get(temp)) {
                            parents.get(temp).add(tmp);
                        }
                        if (!dict.contains(temp))
                            continue;
                        dict.remove(temp);
                        q.offer(temp);
                        levels.put(temp, levels.get(tmp) + 1);
                        List<String> p = new ArrayList<>();
                        p.add(tmp);
                        parents.put(temp, p);
                    }
                    chs[i] = ch;
                }
            }
        }

        if (found) {
            List<String> path = new ArrayList<>();
            path.add(endWord);
            buildPaths(endWord, beginWord, parents, path, result);
        }
        return result;
    }
    private void buildPaths(String word, String beginWord, Map<String, List<String>> parents,
            List<String> path, List<List<String>> result) {
        if (word.equals(beginWord)) {
            List<String> new_path = new ArrayList<>(path);
            Collections.reverse(new_path);
            result.add(new_path);
            return;
        }
        for (String w: parents.get(word)) {
            path.add(w);
            buildPaths(w, beginWord, parents, path, result);
            path.remove(path.size()-1);
        }
    }
}
